include/simppl/any.h

```diff
@@ -1,406 +1,680 @@
+
+    // Unwrap any in any in any in ...
+    std::any anyInner = std::any_cast<Any>(any).value_;
+    if (anyInner.type() == typeid(Any)) {
+      return is_type<std::variant<Types...>>::check(
+          std::any_cast<Any>(any).value_);
+    }
 
-        // FIXME TODO XXX take the message from the codec calls since this is
-        // highly implementation dependent
-        v.set_message_iterator((DBusMessage*)iter.dummy1, iter);
+    std::variant<Types...> result;
+    convert_variant_rec<0, Types...>(std::any_cast<Any>(any), result);
+    return result;
+  }
+};
 
-        dbus_message_iter_next(&orig);
+template <typename Key, typename T, typename Compare, typename Allocator>
+struct as_type<std::map<Key, T, Compare, Allocator>> {
+  static std::map<Key, T, Compare, Allocator> convert(const std::any &any) {
+    assert(any.type() == typeid(IntermediateAnyVec));
+
+    std::map<Key, T, Compare, Allocator> result;
+    for (const std::any &elem : std::any_cast<IntermediateAnyVec>(any).vec) {
+      assert(elem.type() == typeid(IntermediateAnyMapElement));
+      const IntermediateAnyMapElement &mapElem =
+          std::any_cast<IntermediateAnyMapElement>(elem);
+      const Key key = as_type<Key>::convert(mapElem.key);
+      assert(result.find(key) == result.end());
+      result[std::move(key)] = as_type<T>::convert(mapElem.value);
     }
+    return result;
+  }
+};
 
+template <typename T> T Any::as() const {
+  assert(is<T>()); // Ensure the type matches before we convert.
+  return as_type<T>::convert(value_);
+}
 
-    static inline
-    std::ostream& make_type_signature(std::ostream& os)
-    {
-        return os << DBUS_TYPE_VARIANT_AS_STRING;
-    }
+// -------------------------------------DBUS-SIGNATURE-------------------------------------
+
+template <typename T> struct get_signature {
+  static void value(std::ostream &os, const T & /*t*/) {
+    Codec<T>::make_type_signature(os);
+  }
 };
 
-}   // dbus
+template <> struct get_signature<IntermediateAnyVec> {
+  static void value(std::ostream &os, const IntermediateAnyVec &vec) {
+    os << DBUS_TYPE_ARRAY_AS_STRING << vec.elementSignature;
+  }
+};
 
-}   // simppl
+template <> struct get_signature<IntermediateAnyTuple> {
+  static void value(std::ostream &os, const IntermediateAnyTuple &vec) {
+    os << DBUS_STRUCT_BEGIN_CHAR_AS_STRING << vec.elementsSignature
+       << DBUS_STRUCT_END_CHAR_AS_STRING;
+  }
+};
 
+template <> struct get_signature<IntermediateAnyMapElement> {
+  static void value(std::ostream &os,
+                    const IntermediateAnyMapElement &mapElem) {
+    os << DBUS_DICT_ENTRY_BEGIN_CHAR_AS_STRING << mapElem.keyValueSignature
+       << DBUS_DICT_ENTRY_END_CHAR_AS_STRING;
+  }
+};
 
-#endif   // SIMPPL_ANY_H
+template <> struct get_signature<Any> {
+  static void value(std::ostream &os, const Any & /*any*/) {
+    os << DBUS_TYPE_VARIANT;
+  }
+};
 
+template <typename T> std::string get_signature_func(const T &t) {
+  std::ostringstream os;
```

Comment:
Why using a stringstream, when you can use a string?
Appending does not benefit from stringstream in this size of strings.

A std::string would probably more efficient.

Reply:
Because this is how it is already implemented in other parts and I need to interface with them.
I absolutely agree, this is something we should change but in a followup PR?